fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+            }
+        }
+        last_blank_count = ps.skip_blank_block();
+    }
+
+    if let Some(last_comment) = last_comment.take() {
+        body.push(ast::ResourceEntry::Entry(ast::Entry::Comment(last_comment)));
+    }
+    if errors.is_empty() {
+        Ok(ast::Resource { body })
+    } else {
+        Err((ast::Resource { body }, errors))
+    }
+}
+
+fn get_entry<'p>(ps: &mut ParserStream<'p>, entry_start: usize) -> Result<ast::Entry<'p>> {
+    let entry = match ps.source[ps.ptr] {
+        b'#' => ast::Entry::Comment(get_comment(ps)?),
+        b'-' => ast::Entry::Term(get_term(ps, entry_start)?),
+        _ => ast::Entry::Message(get_message(ps, entry_start)?),
+    };
+    Ok(entry)
+}
+
+fn get_message<'p>(ps: &mut ParserStream<'p>, entry_start: usize) -> Result<ast::Message<'p>> {
+    let id = get_identifier(ps)?;
+    ps.skip_blank_inline();
+    ps.expect_byte(b'=')?;
+
+    let pattern = get_pattern(ps)?;
+
+    ps.skip_blank_block();
+
+    let ptr = ps.ptr;
+    let attributes = get_attributes(ps).unwrap_or_else(|_| {
+        ps.ptr = ptr;
+        vec![]
+    });
+
+    if pattern.is_none() && attributes.is_empty() {
+        return error!(
+            ErrorKind::ExpectedMessageField {
+                entry_id: id.name.to_string()
+            },
+            entry_start, ps.ptr
+        );
+    }
+
+    Ok(ast::Message {
+        id,
+        value: pattern,
+        attributes,
+        comment: None,
+    })
+}
+
+fn get_term<'p>(ps: &mut ParserStream<'p>, entry_start: usize) -> Result<ast::Term<'p>> {
+    ps.expect_byte(b'-')?;
+    let id = get_identifier(ps)?;
+    ps.skip_blank_inline();
+    ps.expect_byte(b'=')?;
+    ps.skip_blank_inline();
+
+    let value = get_value(ps)?;
+
+    ps.skip_blank_block();
+
+    let ptr = ps.ptr;
+    let attributes = get_attributes(ps).unwrap_or_else(|_| {
+        ps.ptr = ptr;
+        vec![]
+    });
+
+    if let Some(value) = value {
+        Ok(ast::Term {
+            id,
+            value,
+            attributes,
+            comment: None,
+        })
+    } else {
+        error!(
+            ErrorKind::ExpectedTermField {
+                entry_id: id.name.to_string()
+            },
+            entry_start, ps.ptr
+        )
+    }
+}
+
+fn get_value<'p>(ps: &mut ParserStream<'p>) -> Result<Option<ast::Value<'p>>> {
+    if ps.skip_to_value_start().is_none() {
+        return Ok(None);
+    }
+
+    if ps.is_current_byte(b'{') {
+        let start = ps.ptr;
+        ps.ptr += 1;
+        ps.skip_blank();
+        if ps.is_current_byte(b'*') || ps.is_current_byte(b'[') {
+            let variants = get_variants(ps)?;
+            ps.expect_byte(b'}')?;
+            return Ok(Some(ast::Value::VariantList { variants }));
+        }
+        ps.ptr = start;
+    }
+
+    let pattern = get_pattern(ps)?;
+
+    Ok(pattern.map(ast::Value::Pattern))
+}
+
+fn get_attributes<'p>(ps: &mut ParserStream<'p>) -> Result<Vec<ast::Attribute<'p>>> {
```

Comment:
For now, yes. I'd like to consider improving the recovery here eventually by recognizing that the error originated in the attribute, jumping to the next line that either is a start of new entry or starts with `__"."` and if the latter, resume parsing attributes for that entry.

This would allow us to recover from:

```
key = Value
    .attr = Value { $4 }
    .attr2 = Value 2
```

by salvaging the `attr2` into `key` message. But that's beyond the scope of this patch, so for now, I fixed it :)